Ix.NET/Source/System.Interactive/Retry.cs

```
@@ -45,8 +44,108 @@ public static IEnumerable<TSource> Retry<TSource>(this IEnumerable<TSource> sour
                 throw new ArgumentOutOfRangeException(nameof(retryCount));
             }
 
-            return new[] { source }.Repeat(retryCount)
-                                 .Catch();
+            return RetryFinite(source, retryCount);
+        }
+
+        private static IEnumerable<TSource> RetryInfinite<TSource>(IEnumerable<TSource> source)
+        {
+            while (true)
+            {
+                var enumerator = default(IEnumerator<TSource>);
+                try
+                {
+                    enumerator = source.GetEnumerator() ?? throw new NullReferenceException();
```

Wouldn't it be better to use `if (enumerator == null) continue;` instead of throwing an exception that is catched three lines later?

---

That's supposed to be an uncommon case, besides, `GetEnumerato` can throw on its own so the try-catch should be good enough for both.

---

I wouldn't even check that case, I know there are no real established contracts on the GetEnumerator API but it should be expected not to return null and if all we do is throw a NRE, we might just use enumerator right away and let it blow up.